core/src/main/java/com/scalar/db/common/AbstractDistributedTransactionManager.java

```diff
@@ -66,9 +72,17 @@ public Optional<String> getTable() {
     return tableName;
   }
 
-  protected DistributedTransaction activate(DistributedTransaction transaction)
+  protected DistributedTransaction decorate(DistributedTransaction transaction)
       throws TransactionException {
-    return new StateManagedTransaction(transaction);
+    DistributedTransaction decorated = transaction;
+    for (DistributedTransactionDecorator transactionDecorator : transactionDecorators) {
+      decorated = transactionDecorator.decorate(decorated);
+    }
+    return decorated;
+  }
+
+  public void addTransactionDecorator(DistributedTransactionDecorator transactionDecorator) {
```

Comment:
We can add a transaction decorator with this method. This method is a `public` scope so that we can also add a transaction decorator externally.